src/main/scala/rifs/business/tables/ApplicationTables.scala

```diff
@@ -74,34 +75,43 @@ class ApplicationTables @Inject()(dbConfigProvider: DatabaseConfigProvider)(impl
     (applicationTable returning applicationTable.map(_.id)) += ApplicationRow(None, applicationFormId)
   }
 
+  override def fetchAppWithSection(id: ApplicationId, sectionNumber: Int): Future[Option[(ApplicationRow, Option[ApplicationSectionRow])]] = db.run {
+    appWithSectionC(id, sectionNumber).result.headOption
+  }
+
   override def fetchSection(id: ApplicationId, sectionNumber: Int): Future[Option[ApplicationSectionRow]] = db.run {
     appSectionC(id, sectionNumber).result.headOption
   }
 
   override def fetchSections(id: ApplicationId): Future[Set[ApplicationSectionRow]] = db.run(appSectionsC(id).result).map(_.toSet)
 
   override def saveSection(id: ApplicationId, sectionNumber: Int, answers: JsObject, completedAt: Option[LocalDateTime] = None): Future[Int] = {
-
-    fetchSection(id, sectionNumber).flatMap {
-      case Some(row) =>
-        preCheckSaveAnswers(row.answers, answers, completedAt) match {
-          case true => db.run(appSectionC(id, sectionNumber).update(row.copy(answers = answers, completedAt = completedAt)))
-          case false => Future.successful(1)
-        }
-      case None =>
-        db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+    fetchAppWithSection(id, sectionNumber).flatMap {
+      case Some((app, Some(section))) => areDifferent(section.answers, answers) || completedAt.isDefined match {
+        case true => db.run(appSectionC(id, sectionNumber).update(section.copy(answers = answers, completedAt = completedAt)))
+        case false => Future.successful(1)
+      }
+      case Some((app, None)) => db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+      case None => Future.successful(0)
```

Comment:
That's looks scary to me - we return magic numbers (are these number of rows updated?) upon some error conditions and callers are not checking for these values.